### DISCUSSION TITLE
net: transport abstraction #28165

### DISCUSSION THREAD

**Discussion by rajarshimaitra** [Bitshala/BitcoinCore-PR-Review-Club#45] [PR Reviews]
### Session Details
- Date: 07-09-2023
- Time: IST 20:00 (UTC 14:30)
- PR: https://github.com/bitcoin/bitcoin/pull/28165
- Context: `[net][p2p]`
- Difficulty: Moderate
- Language: `[c++]`
- Notes and Questions: Coming Soon

### Summary
This PR furthers the P2P message serialization/deserialization abstraction introduced in https://github.com/bitcoin/bitcoin/pull/16202 and https://github.com/bitcoin/bitcoin/pull/16562, in preparation for teaching the BIP324 v2 transport (making this part of https://github.com/bitcoin/bitcoin/issues/27634). However, nothing in this PR is BIP324-specific, and it contains a number of independently useful improvements.

### Learning
 - P2P message handling.
 - in-depth into netprocessing.cpp.
> Note: Drop your PR and review related questions below

**Comment by stratospher** [comment]
# Notes
* Bitcoin nodes communicate by exchanging lots of different P2P messages. P2P messages get converted into wire bytes which are then sent over.
* P2P messages -> wire bytes and wire bytes -> P2P messages conversion is now handled by a single `Transport` class which does both sending and receiving.
* Currently, the P2P messages are unencrypted and any passive observer with access to the wire bytes can figure out the P2P messages exchanged between nodes. We'll refer to this P2P transport protocol as v1 P2P here.
* v1 P2P starts with version handshake where the initiator node sends a "version" P2P message and expects a "verack" P2P message back.
* wire bytes format for V1 P2P message:
![Screenshot from 2023-09-07 13-17-53](https://github.com/Bitshala/Bitcoin-PR-Review-Club/assets/44024636/32e6ffe5-a7fa-4672-b3a0-6f331bf360cf)

* BIP 324 introduces a new P2P protocol which can send encrypted P2P messages. We'll refer to this P2P transport protocol as v2 P2P here.
* wire bytes format for V2 P2P message:
![Screenshot from 2023-09-07 13-18-17](https://github.com/Bitshala/Bitcoin-PR-Review-Club/assets/44024636/8bc80976-2f56-4f1b-b2ed-202330eee182)

* v2 P2P starts with a DH key negotiation phase to compute the encryption keys and then carries on with the version handshake.
* This PR merges the `TransportDeserializer`(receiving side) and `TransportSerializer`(sending side) classes into a single Transport class in preparation for V2Transport. This is done because sending side and receiving side are dependent on each other during the DH key negotiation in V2Transport.

# Questions
1. Did you review the PR? Concept ACK, approach ACK, tested ACK, or NACK? What was your review approach?
2. What are the different net threads? what do they do?
    - HINT: https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/doc/developer-notes.md#threads
![IMG-0040](https://github.com/Bitshala/Bitcoin-PR-Review-Club/assets/44024636/f3322075-5fe4-47b8-b675-5b35171b6df5)
3. What's the difference between net and net processing? HINT: notes in https://bitcoincore.reviews/20721
4. How has the sending logic changed for v1 P2P?
    - HINT: in [`SocketSendData()`](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L913-L983), notice usage of [`SetMessageToSend`](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L924), [`GetBytesToSend`](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L930) and [`MarkBytesSent`](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L955). Previously, [`SocketSendData()`](https://github.com/bitcoin/bitcoin/blob/0a55bcd2996d878a831e5ae36b7e6a869b0dda2b/src/net.cpp#L836-L892) did not use these. What advantages do using these new functions provide?
    - V1 message checksum calculation now happens in [`SetMessageToSend`](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L924) guarded by `cs_vSend` lock at it's call site in [`SocketSendData()`](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L924). Previously, it happened in [`prepareForTransport()`](https://github.com/bitcoin/bitcoin/blob/0a55bcd2996d878a831e5ae36b7e6a869b0dda2b/src/net.cpp#L822-L834) in [`PushMessage()`](https://github.com/bitcoin/bitcoin/blob/0a55bcd2996d878a831e5ae36b7e6a869b0dda2b/src/net.cpp#L2911) which wasn't guarded by this lock.
    -  what are optimistic/non-optimistic sends?
6. what does `PushMessage()` and `SocketSendData()` do? where does the actual sending happen?
    - HINT: `PushMessage()` [before](https://github.com/bitcoin/bitcoin/blob/0a55bcd2996d878a831e5ae36b7e6a869b0dda2b/src/net.cpp#L2891-L2932) vs [now](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L2988-L3026), `SocketSendData()` [before](https://github.com/bitcoin/bitcoin/blob/0a55bcd2996d878a831e5ae36b7e6a869b0dda2b/src/net.cpp#L836-L892) vs [now](https://github.com/bitcoin/bitcoin/blob/d2ccca253f4294b8f480e1d192913f4985a1af08/src/net.cpp#L913-L983)

**Comment by stratospher** [comment]
call graph for `SocketHandlerConnected()` in net thread:

| before  | now |
| ------------- | ------------- |
| ![IMG-0039](https://github.com/Bitshala/Bitcoin-PR-Review-Club/assets/44024636/7735370e-ff5e-44c7-bdad-e4d35f398662) | ![IMG-0038](https://github.com/Bitshala/Bitcoin-PR-Review-Club/assets/44024636/d5261b20-05c4-4d69-9554-bbb39183872e) |
